### ISSUE TITLE
Assigning incompatible types does not raise an error

### ISSUE THREAD

**Issue by bagbag** [microsoft/TypeScript#33551]
**TypeScript Version:** tested with 3.5, 3.6, next

**Search Terms:**
typescript allows assigning incompatible types
typescript allows assigning generic interfaces without full overlap

**Code**

```ts
type Entity = { id: string };
type Participant = Entity & { participantName: string };
type Trainer = Entity & { trainerName: string };
type Master = Trainer & { masterName: string };

interface EntityRepository<T extends Entity> {
  load<U extends T = T>(id: string): Promise<U>;
}

// should not work as Participant doesn't extend Trainer
const a: EntityRepository<Trainer> = (null as any) as EntityRepository<Participant>;

// doesn't work as expected, as Participant doesn't extend Trainer
a.load<Participant>("");

// works as expected, as Master extends Trainer
a.load<Master>("");
```

**Expected behavior:**
Assigning `EntityRepository<Participant>` to `EntityRepository<Trainer>` should raise an error, as `Participant` is *not* assignable to `Trainer`.

**Actual behavior:**
Typescript does not complain at all for the assignation.

**Extra note:**
If I remove `U extends T = T` from `load` and return `Promise<T>` instead of `Promise<U>` everything works as expected.
A workaround I found for myself is adding a fake property `_type: T` to the interface.

**Playground Link:**
[Playground](https://www.typescriptlang.org/play/index.html?ts=Nightly#code/C4TwDgpgBAogdsAlqKBeKBvKiAmAuKAZ2ACdE4BzKAXwG4AoUSKABQEMSkBjRMNhNLATIQUAGSYofToh58EAOTYBbCAWJlKNBk2gAVEm3IQSg+EhQSspI3BNLV60uSp1G4aAFk2xE4IO2flZQyj7A9ipqRM5abvTk4SQAZmxc0OYiAEoQYAD2hMi5JCAAPHpQEAAe4XA4hEIWIAB8mPRQUAA2uWw4JQCqFdUQtfXl6HpNABS4TpoUAJQELCS5yoiEEP1NDNT09AD0+0QAFrkArh04UHC5wFAA7kUA1lA+rBzcvPx3OLkQhHAAOR3Ko1K4BYwkehcXJwYivAgZUDZPIFYBFUoQuwkFroSZwC4dV71fggebEhpZHL5QrFErsGRyb7bPaHKC-f5Au6PEgvN5VSBccI4AA0FIZn3kPz+AOBgzBUCxJnobAAdF0evSPrIvggpgAifXzBgHI48p4k+oCiBCiCiineXymUHDOqKwyQlXq7q9R2JA1GhhAA)

**Related Issues:**
[https://github.com/Microsoft/TypeScript/issues/31006](https://github.com/Microsoft/TypeScript/issues/31006)

**Comment by jack-williams** [comment]
This is fixed by #31029.

**Comment by bagbag** [comment]
Oh, I didn't see the pull request in the related issue and wasn't sure whether it is the same bug.

Thanks!
